File: lib/src/embedded/boundings/camera/application/codec.dart
part of mapbox_navigation_flutter;

class CameraAPICodec extends StandardMessageCodec {
  const CameraAPICodec();
  @override
  void writeValue(WriteBuffer buffer, Object? value) {
    if (value is CameraOptions) {
      buffer.putUint8(128);
      writeValue(buffer, value.encode());
    } else if (value is MapAnimationOptions) {
      buffer.putUint8(129);
      writeValue(buffer, value.encode());
    } else if (value is MbxEdgeInsets) {
      buffer.putUint8(130);
      writeValue(buffer, value.encode());
    } else if (value is ScreenCoordinate) {
      buffer.putUint8(131);
      writeValue(buffer, value.encode());
    } else if (value is ScreenCoordinate) {
      buffer.putUint8(132);
      writeValue(buffer, value.encode());
    } else if (value is CameraState) {
      buffer.putUint8(133);
      writeValue(buffer, value.encode());
    } else {
      super.writeValue(buffer, value);
    }
  }

  @override
  Object? readValueOfType(int type, ReadBuffer buffer) {
    switch (type) {
      case 128:
        return CameraOptions.decode(readValue(buffer)!);
      case 129:
        return MapAnimationOptions.decode(readValue(buffer)!);
      case 130:
        return MbxEdgeInsets.decode(readValue(buffer)!);
      case 131:
        return ScreenCoordinate.decode(readValue(buffer)!);
      case 132:
        return ScreenCoordinate.decode(readValue(buffer)!);
      case 133:
        return CameraState.decode(readValue(buffer)!);
      default:
        return super.readValueOfType(type, buffer);
    }
  }
}

File: android/src/main/kotlin/com/eopeter/fluttermapboxnavigation/boundings/camera/application/Codec.kt
package com.eopeter.fluttermapboxnavigation.boundings.camera.application

import com.eopeter.fluttermapboxnavigation.core.*
import io.flutter.plugin.common.StandardMessageCodec
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer

@Suppress("UNCHECKED_CAST")
object CameraApiCodec : StandardMessageCodec() {
    override fun readValueOfType(type: Byte, buffer: ByteBuffer): Any? {
        return when (type) {
            128.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    CameraOptions.fromList(it)
                }
            }
            129.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    MapAnimationOptions.fromList(it)
                }
            }
            130.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    MbxEdgeInsets.fromList(it)
                }
            }
            131.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    ScreenCoordinate.fromList(it)
                }
            }
            132.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    CameraState.fromList(it)
                }
            }
            133.toByte() -> {
                return (readValue(buffer) as? List<Any?>)?.let {
                    CoordinateBounds.fromList(it)
                }
            }
            else -> super.readValueOfType(type, buffer)
        }
    }
    override fun writeValue(stream: ByteArrayOutputStream, value: Any?) {
        when (value) {
            is CameraOptions -> {
                stream.write(128)
                writeValue(stream, value.toList())
            }
            is MapAnimationOptions -> {
                stream.write(129)
                writeValue(stream, value.toList())
            }
            is MbxEdgeInsets -> {
                stream.write(130)
                writeValue(stream, value.toList())
            }
            is ScreenCoordinate -> {
                stream.write(131)
                writeValue(stream, value.toList())
            }
            is CameraState -> {
                stream.write(132)
                writeValue(stream, value.toList())
            }
            is CoordinateBounds -> {
                stream.write(133)
                writeValue(stream, value.toList())
            }
            else -> super.writeValue(stream, value)
        }
    }
}